### PR TITLE
Resolves #1219, Implement File.extname method

### LINKED ISSUE #1219: Implement File.extname method
In Asciidoctor.js we are using this implementation:

```ruby
  def self.extname(path)
    return '' if path.nil_or_empty?
    last_dot_idx = path[1..-1].rindex('.')
    last_dot_idx.nil? ? '' : path[(last_dot_idx + 1)..-1]
  end
```

### PR DESCRIPTION
**Pull request by Mogztter**
Resolves #1219

### CODE REVIEW

**Comment by elia** [comment]
@Mogztter can you add a spec for this? (either from ruby spec or in spec/opal)

File: opal/corelib/file.rb [hunk 1]
```diff
@@ -30,6 +30,12 @@ def basename(path)
       split(path)[-1]
     end
 
+    def extname(path)
+      return '' if path.nil_or_empty?
```
Comment (elia) [inline opal/corelib/file.rb]: Is `#nil_or_empty?` defined somewhere?

**Comment by Mogztter** [comment]
@elia Sure. I've never done that before, I just need to add a new line in here https://github.com/opal/opal/blob/master/spec/rubyspecs#L129 ? or write a spec file here: https://github.com/opal/opal/tree/master/spec/opal/core ?

File: opal/corelib/file.rb (hunk 1 above)
Comment (Mogztter) [inline opal/corelib/file.rb]: Good catch, I though this was a standard Ruby method but in fact this is an Asciidoctor core extension [nil_or_empty.rb](https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/core_ext/object/nil_or_empty.rb).
Do you want me to add `nil_or_empty` extension or replace the line with a standard Ruby code ?

Comment (vais) [inline opal/corelib/file.rb]: My :+1: for *standard Ruby code* :smile:

Comment (elia) [inline opal/corelib/file.rb]: Indeed only stuff from corelib is allowed in corelib 😊

**Comment by elia** [comment]
@mogztter see HACKING.md for an explanation on how to deal with specs

(excuse brevity, AFK, I'll get into more details as soon as I can 🙂)

**Comment by Mogztter** [comment]
@elia I'm now 100% compatible with Ruby spec !

File: opal/corelib/file.rb [hunk 2]
```diff
@@ -30,6 +30,18 @@ def basename(path)
       split(path)[-1]
     end
 
+    def extname(path)
+      raise TypeError, 'no implicit conversion of nil into String' if path.nil?
+      path = path.to_path if path.respond_to?(:to_path)
+      raise TypeError, "no implicit conversion of #{path.class} into String" unless path.is_a?(String)
+      filename = basename(path)
```
Comment (Mogztter) [inline opal/corelib/file.rb]: `File.basename` method is not fully compliant with the Ruby specs but all tests passed nevertheless :wink:

File: opal/corelib/file.rb [hunk 3]
```diff
@@ -30,6 +30,18 @@ def basename(path)
...
+      return '' if filename.nil?
+      last_dot_idx = filename[1..-1].rindex('.')
+      extname = last_dot_idx.nil? ? '' : filename[(last_dot_idx + 1)..-1]
+      return '' if extname.tr('.', '').empty? # extname contains only dots
```
Comment (Mogztter) [inline opal/corelib/file.rb]: I tried to use a Regexp `extname.match(/^.+$/)` but didn't make it work...

Comment (elia) [inline opal/corelib/file.rb]: Have you tried escaping the dot? `\.`

File: spec/rubyspecs [hunk 4]
```diff
@@ -127,3 +127,5 @@ rubyspec/library/pp
 rubyspec/library/set
 rubyspec/library/singleton
 rubyspec/library/stringscanner
+
+rubyspec/core/file/extname_spec
```
Comment (elia) [inline spec/rubyspecs]: Can you move this up to keep the loose alphabetical order?

**Comment by elia** [comment]
> @elia I'm now 100% compatible with Ruby spec !

Great!

File: opal/corelib/file.rb (hunk 3 above)
Comment (Mogztter) [inline opal/corelib/file.rb]: Yes I did. 
I've took a different approach, now I'm just checking the last dot index with the filename length.

File: spec/rubyspecs (hunk 4 above)
Comment (Mogztter) [inline spec/rubyspecs]: Ok

**Comment by elia** [comment]
:+1:

**Comment by Mogztter** [comment]
Cool, thanks :tada:

**Pull request merged**